### ISSUE TITLE
What are the possible labels?

### ISSUE THREAD

**Issue by Ironholds** [openvenues/libpostal#48]
As someone trying to fit the output into a tabular datastructure it'd be good to know what the range of labels for parsed addresses are (I've tried to dig through the code, but...there's kind of a lot of it!)

**Comment by Ironholds** [comment]
Okay, I thiiink I've answered my own question by finding [this](https://github.com/openvenues/libpostal/blob/da62ff309eb5eb20476f48c7098b2d9eac32c583/src/address_parser.h#L99-L108). Are those the fields? If so I'll refactor the issue to "I should submit a pull request to improve the documentation a bit"

**Comment by thatdatabaseguy** [comment]
Hey Oliver - documentation is << code indeed :-). A general documentation ticket would be welcome, and is something on my mind of late. I'll be publishing automated training scripts shortly with some accompanying docs. That should make it slightly more clear how everything fits together, though it's an ongoing project.

The current list of parser tags can be found in [this Python module](https://github.com/openvenues/libpostal/blob/master/scripts/geodata/address_formatting/formatter.py) as well as address_parser.h, as you mentioned. They're more-or-less inherited from OpenCage's [address-formatting](https://github.com/OpenCageData/address-formatting) repo, and are open for discussion. We can version the models for the next release if there are any breaking changes.

Here are brief descriptions of each tag:

- **house**: venue/facility/building/house name
- **house_number**: the street-facing building number if there is one (addr:housenumber in OSM)
- **road**: street name (addr:street in OSM)
- **suburb**: usually a neighborhood or unofficial sub-city place name
- **city_district**: more official subdivision of a city, like a borough in London or NYC (in NYC boroughs are also counties, so the label for "Brooklyn" for instance may vary depending on context).
- **city**: city/town/village/hamlet
- **state_district**: in the US, UK and Ireland, state_district corresponds to a county, but it varies by country.
- **state**: state/province/region/admin1 (England/Scotland/Wales/Northern Ireland are technically considered "state boundaries" 
- **postcode**: postal code where available
- **country**: sovereign states and dependent territories (anything with an [ISO 3166-1 code](https://en.wikipedia.org/wiki/ISO_3166-1#Current_codes))

There are per-country mappings of [OpenStreetMap admin boundaries](http://wiki.openstreetmap.org/wiki/Tag:boundary%3Dadministrative) to our tags [here](https://github.com/openvenues/libpostal/tree/master/resources/boundaries/osm).

For the next release I'm adding a few more labels based on the use cases and data sets encountered so far. The current candidates are:

- **level**: floor number
- **building**: building within a larger apartment complex/housing block
- **unit**: apartment/flat/unit number
- **pobox**: for post office addresses
- **category**: for parsing geocoder input like "restaurants in Fort Greene, Brooklyn"

Technically the labels are just strings so if folks want to train on a different data set, proprietary or otherwise, they can add their own tags and retrain without affecting libpostal, though I don't expect it to happen often.

Re: tabular data, one thing to note is that it is possible, though unlikely, for there to be > 1 instance of the same field. The parser is a greedy Markov-style sequence model, making a prediction at each token and lumping them into phrases at the end. Usually if you do encounter duplicate fields, it's a parse error, but for some input it can be legitimate (city, state, postcode was copy-pasted twice, some of the addresses on [this list](https://www.mjt.me.uk/posts/falsehoods-programmers-believe-about-addresses/), etc.)

**Comment by Ironholds** [comment]
Ooh, interesting. Thanks for the detailed reply! I'll noodle on doc work. Question; in the event of duplicate fields, then as a general rule should I trust the first entry, the last entry, is it non-deterministic? Obviously I'll pick the wrong one some of the time :)

**Comment by thatdatabaseguy** [comment]
Depends on the input and the language. I'd guess the first entry in most cases, but safer to store both if possible.

Also, somewhat related, rumor has it you might be working on an R binding. Is that the case and are any libpostal changes needed to make life easier?

**Comment by Ironholds** [comment]
Who told you that?!

\>.>

Iff I was working on an R binding it would be for personal use right now rather than immediate release. So, you know. Some unknown amount of time before it came out. Hopefully not *much* time but some amount of time. Not that I'm saying I am working on an R binding.

But, in this hypothetical situation, it's a complete unknown as to how CRAN would react to a binding for the simple reason that while pkg-config allows the inclusion of an installed libpostal instance into an Rcpp project, actually *wrapping* the library would be a massive pain; 5GB of free space in a specified location and win-builder go together like chalk and cheese. So something along the lines of rgeolocate, where all the dependencies live inside the R package, ech, probably hard-to-impossible to produce in an acceptable form.

One sort-of-way around that _maybe_ would be if libpostal could be passed an install flag that said "don't actually allocate those five gigs, just fall back to a predefined set of calls and responses for this install", but even there actually manipulating that and detecting whether things should be installed with or without that flags might be thorny. Plus, CRAN's C version is ancient :|.

An alternative would be if someone familiar with libpostal's install process and defaults were to mock up what a configure.ac function for detecting the presence of libpostal on a unix system would look like, so that any binding could at least recognise that dependencies weren't met in a user-friendly way. In my experience CRAN tends to be displeased by that but not so displeased they block package release.

Obviously, though, this is all hypothetical.

**Comment by thatdatabaseguy** [comment]
Educated guess :-), saw [Mr. Wickham's tweet](https://twitter.com/hadleywickham/status/703035174234120192) when this launched. I'll file it as a "we can neither confirm nor deny..."

Do all CRAN packages have to build on Windows? Reports so far have been that libpostal won't even compile on Windows, and having no such machine to test on, I haven't looked into it yet. The use of mmap in our [sparkey](https://github.com/spotify/sparkey) dependency appears to be one of the culprits, so some ifdef-ery may be necessary.

One example I've seen of the second approach you describe is [the Rstats libxml2 binding](https://github.com/hadley/xml2/blob/master/configure). That uses pkg-config and expects the C lib to be installed separately, though it's a much more ubiquitous C lib that can be installed through package managers.

Publishing libpostal binary packages on Github (sans data files) might make the separate install more palatable, and would be useful for moving toward one-line installs in other package managers (Python's PyPI, Node's npm, etc). That would mean the data directory gets specified in a runtime config rather than at configure/compile-time, though to do anything useful with libpostal the models would need to be downloaded and accessible. I haven't seen many open-source packages that publish production models, so kind of making it up as I go.

For the autoconf script, I wrote something similar in the [PHP binding](https://github.com/openvenues/php-postal/blob/master/config.m4) that can be reused with a few minor modifications.

Also, re: size, libpostal should only take up ~2.2G all said and done. The only thing I can think of that would account for a 5G install is if you changed the datadir at some point and re-ran ```make```, causing a second download to the new location. Otherwise should be significantly smaller.

**Comment by Ironholds** [comment]
They don't; it limits the portability, of course, but it's not a hard requirement (you simply note platform: unix in the DESCRIPTION). Hadley's libxml2 is a good example; so is the (now deprecated, thank god) RCurl package, so is RJava (not deprecated, augh).

Thanks for the PHP pointer! Apologies on my mistake vis a vis the size; I suspect the same problem would occur whenever the size starts with a G :). They don't even like packages to be that big. Publishing on GH, or debianising, would definitely help there, yeah.

**Comment by Rickasaurus** [comment]
It would also be nice to have support for variations on PO BOX and C/O. Right now the PO Boxes in particular are really throwing it:

for "PO BOX 2143 New York NY"
"house_number:" "po",
"road": "box",
"postcode:" "2143",
"city": "new york",
"state": ny

**Comment by thatdatabaseguy** [comment]
Indeed, there are several improvements being worked on in the [parser-data](https://github.com/openvenues/libpostal/tree/parser-data) branch at the moment. po_box is among the new labels for the next release.

The PO boxes we train on will be randomly generated, but it doesn't matter as libpostal only really cares about the number of digits in this case. Some PO boxes use letters or numbers + letters, which we can also append randomly.

Addresses with c/o might be slightly more difficult as there's not much in the way of training data, though the list you posted from the IRS in #44 is a good start. There are gazetteers of first and last names from Wikipedia/Wiktionary included in libpostal's dictionaries, which could at least help in identifying Western names following a known phrase. Not sure yet if that'll make it into the release but it does come up in companies, [most German addresses](http://www.german-way.com/germans-we-dont-need-apartment-numbers/), universities, etc.

**Comment by Rickasaurus** [comment]
That's great to hear. I'm very glad you're committed to constant improvements.

We have a bunch of C/O data but I'm not sure if our license would allow us to share it. I'll look into it.

**Comment by Rickasaurus** [comment]
The address data in the panama papers dataset has a ton of interesting structure including PO BOXes and Apartment/Flat numbers: https://offshoreleaks.icij.org/pages/database

**Comment by thatdatabaseguy** [comment]
👍 these are great.

They're not quite in a form that's directly usable as training data, but having raw sub-building addresses is helpful for estimating prior probabilities of certain patterns ("1st Floor" vs. "1/F" vs. "Floor 1", etc.) and discovering new ones.

In the next release I'm working on, data sets like this one help inform the new per-language/country address configs ([English](https://github.com/openvenues/libpostal/blob/parser-data/resources/addresses/en.yaml) and [Spanish](https://github.com/openvenues/libpostal/blob/parser-data/resources/addresses/es.yaml) are implemented so far). Said configs tell libpostal how to randomly generate some of the components we don't have in geocoded training addresses including apartment/floor numbers, PO boxes, etc. in a localized way.

**Comment by thatdatabaseguy** [comment]
@Ironholds sent you a pull request for the new labels in 1.0.

@Rickasaurus PO boxes are now implemented in a number of languages. C/O is probably not going to be implemented for the time being, not enough data to learn it well.
